test/control_dependencies.cpp

```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+* Copyright 2017-2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <algorithm>
+#include <cstdio>
+#include <iostream>
+#include <list>
+#include <memory>
+
+#include "gtest/gtest.h"
+#include "ngraph/autodiff/adjoints.hpp"
+#include "ngraph/file_util.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/ngraph.hpp"
+#include "ngraph/op/batch_norm.hpp"
+#include "ngraph/op/get_output_element.hpp"
+#include "ngraph/op/parameter.hpp"
+#include "ngraph/pass/manager.hpp"
+#include "ngraph/pass/visualize_tree.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/runtime/cpu/pass/cpu_fusion.hpp"
+#include "ngraph/serializer.hpp"
+#include "ngraph/util.hpp"
+#include "nlohmann/json.hpp"
+#include "util/all_close.hpp"
+#include "util/autodiff/backprop_function.hpp"
+#include "util/autodiff/numeric_compare.hpp"
+#include "util/ndarray.hpp"
+#include "util/random.hpp"
+#include "util/test_tools.hpp"
+
+using namespace ngraph;
+using namespace std;
+
+class ControlDependencyOp : public ngraph::op::Op
+{
+public:
+    virtual std::shared_ptr<Node> copy_with_new_args(const NodeVector& new_args) const override
+    {
+        auto clone = make_shared<ControlDependencyOp>(new_args, std::set<std::shared_ptr<Node>>{});
+        return clone;
+    }
+
+    ControlDependencyOp(const NodeVector& args, const std::set<std::shared_ptr<Node>>& deps)
+        : Op("ControlDependencyOp", args)
+    {
+        if (args.size() == 0 && deps.size() == 0)
+        {
+            throw ngraph_error("Expected some arguments or dependencies");
+        }
+
+        if (deps.size() != 0)
+        {
+            m_control_dependencies.insert(deps.begin(), deps.end());
+        }
+
+        if (args.size() != 0)
+        {
+            set_output_type(0, args.at(0)->get_element_type(), args.at(0)->get_shape());
+        }
+        else
+        {
+            auto dn = *(deps.begin());
+            set_output_type(0, dn->get_element_type(), dn->get_shape());
+        }
+    }
+};
+
```

Comment:
Check that something terrible doesn't happen if you introduce a circular dependency with control deps, something which can happen easier than you might think.